pkg/addonfactory/factory.go

```diff
@@ -0,0 +1,91 @@
+package addonfactory
+
+import (
+	"embed"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+)
+
+const AddonDefaultInstallNamespace = "open-cluster-management-agent-addon"
+
+type AgentAddonFactory struct {
+	scheme            *runtime.Scheme
+	fs                embed.FS
+	templateFiles     []string
+	templateConfig    interface{}
+	agentAddonOptions agent.AgentAddonOptions
+}
+
+func NewAgentAddonFactoryWithTemplates(addonName string, fs embed.FS, templateFiles ...string) *AgentAddonFactory {
+	return &AgentAddonFactory{
+		fs:            fs,
+		templateFiles: templateFiles,
+		agentAddonOptions: agent.AgentAddonOptions{
+			AddonName:       addonName,
+			Registration:    nil,
+			InstallStrategy: nil,
+		},
+	}
+}
+
+// WithScheme is an optional configuration, only used when the manifests have customized resource types.
+func (f *AgentAddonFactory) WithScheme(scheme *runtime.Scheme) *AgentAddonFactory {
+	f.scheme = scheme
+	return f
+}
+
+// WithTemplateConfig adds the template config struct.
+// the reserved fields (ClusterName,AddonInstallNamespace and KubeConfigSecret) are not mandatory in the config.
+func (f *AgentAddonFactory) WithTemplateConfig(config interface{}) *AgentAddonFactory {
+	f.templateConfig = config
+	return f
+}
+
+// WithInstallStrategy defines the installation strategy of the manifests prescribed by Manifests(..).
+// installType is the type of the installation strategy. currently only supported values: "agent.InstallAll".
+// installNamespace is target deploying namespace in the managed cluster upon automatic addon installation.
+func (f *AgentAddonFactory) WithInstallStrategy(installType agent.StrategyType, installNamespace string) *AgentAddonFactory {
```

Comment:
maybe set InstallStrategy, otherwise it will be hard to evolve

Reply:
done.